source/Irrlicht/CImageLoaderTGA.cpp

```diff
@@ -138,7 +139,8 @@ IImage* CImageLoaderTGA::loadImage(io::IReadFile* file) const
 	if (header.ColorMapType)
 	{
 		// Create 32 bit palette
-		const irr::u16 paletteSize = core::max_((u16)256, header.ColorMapLength);	// ColorMapLength can lie, but so far we only use palette for 8-bit, so ensure it has 256 entries
+		// `core::max_()` is not used here because it takes its inputs as references. Since `header` is packed, use the macro `MAX()` instead:
```

Comment:
what about `std::max`?

Reply:
> what about `std::max`?

Same issue, see https://github.com/minetest/irrlicht/pull/288#discussion_r1485651314. Since `std::max` is technically a function, it seems the undefined behavior would technically persist as we pass on a reference to an unaligned struct member.

The macro is one solution, another would be to first copy `header.ColorMapLength` to a local variable. I can live with either.

Reply:
That's unfortunate.
Now since this seems like such a common trap I am left wondering: is it *actually* UB if the template never takes the address of the argument?

Reply:
That's a good point, but I think we can not make sure of that. As far as I know, references always make use of pointers in the background.

Reply:
> As far as I know, references always make use of pointers in the background.

I don't think they have to; the C++ compiler will most probably optimize references out in many cases on higher optimization levels, especially const references to types which are always cheaper to pass by value (as is the case here) if it can inline the function (as is the case for `std::max`).

But you're right that we most probably can't rely on this, especially as the compiler needn't optimize on `-O0`.